### ISSUE TITLE
Encourage using IDictionary interface for parameters

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#13852]
### Encourage developers to use the [`IDictionary`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.idictionary-2) interface for the parameter type
(if the `[Hashtable]` type is not explicitly required which I think it almost always the case)

Currently, In most custom functions, the specific `[Hashtable]` dictionary type is used for parameter type where it would be better to use the more generic `IDictionary` interface so that is would accept any `ICollection` or `IEnumerable` (as e.g. the `[Ordered]` type), knowing that the enumeration of all `IDictionary` Objects are common and most native PowerShell cmdlet accept the `IDictionary` interface.
The reason that specific `[Hashtable]` dictionary type is used is because the `[Hashtable]` type is more known (and default) within PowerShell and has simply a shorter type name.

### Proposed technical implementation details (optional)
* Add the `IDictionary` interface to the default list of type accelerators:

```PowerShell
$TypeAccelerators = [psobject].Assembly.GetType('System.Management.Automation.TypeAccelerators')
$TypeAccelerators::Add('IDictionary', [System.Collections.IDictionary]) # Or just 'Dictionary'?
```
So that a developer/user can use any dictionary object for input:
```PowerShell
Function Test([IDictionary]$Dictionary) {
    [pscustomobject]$Dictionary
}
Test ([Ordered]@{ a = 1; b = 2; c = 3 })
```
Possibly also:
* Add the recommendation to use the `IDictionary` interface to the best practice documentation
* Create a PSScriptAnalyzer  warning when an explicit `[Hashtable]` type is used for a parameter?

Related:
* Feature Request: add [Ordered] type accelerator  #10513
* Enhance hash table syntax #13817

**Comment by vexx32** [comment]
I think you mean `IDictionary`? 🤔

**Comment by SeeminglyScience** [comment]
Currently the engine isn't very smart about casting *to* interfaces.  It's not that big of a deal for `IDictionary`, but before interfaces are recommended that should probably be solved.

This is more often a problem when it comes to `IEnumerable<T>` and `IList<T>`. PowerShell knows how to cast `object[]` to `int[]`, and it knows that `int[]` is `IEnumerable<int>`, but it doesn't know how to cast `object[]` to `IEnumerable<int>`.  Admittedly, that's not an easy problem to solve, but I do think it's worth it.

**Comment by mklement0** [comment]
I think it's fine to already start recommending use of `IDictionary` specifically, given that it is both a common use case and works as intended - though better support such as `object[]` to `IEnumerable<int>` casts would certainly be nice.

I think the challenges of interface use in PowerShell are primarily _conceptual_:

* User usually don't have to deal with interfaces directly, given that PowerShell surfaces even explicit interface implementations as direct object members, and less experienced users may even be unaware of the concept of interfaces altogether.

  * This means that when they see interface-based parameter declarations (e.g., `IDictionary` for `New-Object`'s `-Property` parameter), they need to either rely on the parameter _description_ to know what concrete types to pass or rely on their own advanced .NET knowledge, including knowing how to inspect a given type (instance) for implementing the target interface.

* Script authors, especially those with a C# background, need to be aware that PowerShell generally _ignores_ casts to interfaces and base classes and that such casts matter only for .NET method overload resolution and during PowerShell function/script parameter binding, but even in the latter case what the script / function sees as the parameter _variable_ is the _original type_.

  * In other words: in PowerShell, you never interact with an interface _directly_, only with an implementing type's implementations of its members, which means that inside a script with an interface-typed parameter you need to be aware of what members are truly the interface's, as only they can be relied on.

In addition to documentation, the following would also help:

The ability to inspect a given instance or type for the interfaces it implements as well as their members - see #13865

**Comment by IISResetMe** [comment]
WG-Engine discussed this yesterday, and considered the following points proposed:

  - Labeling and recommending the use of `[IDictionary]` as a type constraint in place of `[hashtable]` in command parameter blocks a "Best Practice" in the documentation
  - Adding a default `[IDictionary]` shorthand to the default list of type accelerators.

## Best practice recommendation (not endorsed)

While we agree `[IDictionary]` is useful as a type constraint for functions or scripts that utilize only the subset of functionality defined by said interface, we disagree its recommendation should qualify as a "best practice" endorsement. 

The majority of use cases involving dictionary-like data structures as parameter arguments in PowerShell can be (and are) successfully implemented using the `[hashtable]` type, and the built-in hashtable literal syntax `@{<keyN>=<valueN>}` is deliberately instrumented to fit the expectations of even novice users (case-insensitive string comparison for keys, least constraint on value type). As a result, in use cases where the ordering of the entries or the identity of the dictionary itself is not important, `[hashtable]` is a much more appropriate as a parameter type constraint - something we'd be contradicting if we started recommending `[IDictionary]` as a "best practice" catch-all parameter constraint for dictionary-like arguments.

As such, it would also be inappropriate to have PSScriptAnalyzer emit warnings for use of concrete dictionary types as parameter constraints - as many perfectly valid use cases exist.

## `[IDictionary]` type accelerator (not endorsed)

Given the limited applicability of the pattern (as described above), we also cannot justify adding `[IDictionary]` to the default list of type accelerators. 
The type literal can be shortened in-place using the `using namespace` directive:

```powershell
using namespace System.Collections

function f {
  param(
    [IDictionary]$Dictionary
  )
}
```

Although we cannot endorse the original proposal to recommend it as "best practice", we are more than happy to see, and review, proposals for documentation that explain the `[IDictionary]` pattern more in depth, perhaps in `about_Hash_Tables` :-)
